File: backend/app/services/priorities.py
from datetime import date

from app.models import AppointmentStatus, AppointmentType, Priority
from app.schemas import Appointment, PriorityPlanItem


_PRIORITY_WEIGHTS = {
    Priority.BAIXA: 10,
    Priority.MEDIA: 30,
    Priority.ALTA: 60,
}

_TYPE_WEIGHTS = {
    AppointmentType.PROVA: 25,
    AppointmentType.TRABALHO: 20,
    AppointmentType.TAREFA: 15,
    AppointmentType.APRESENTACAO: 18,
    AppointmentType.LEITURA: 8,
}

_STATUS_WEIGHTS = {
    AppointmentStatus.PENDENTE: 20,
    AppointmentStatus.EM_ANDAMENTO: 10,
    AppointmentStatus.CONCLUIDO: -40,
}


def _deadline_weight(appointment_date: date, reference_date: date) -> int:
    delta_days = (appointment_date - reference_date).days

    if delta_days < 0:
        return 35
    if delta_days == 0:
        return 40
    if delta_days <= 2:
        return 35
    if delta_days <= 7:
        return 25
    if delta_days <= 14:
        return 15
    return 5


def calculate_priority_score(
    appointment: Appointment,
    reference_date: date | None = None,
) -> int:
    today = reference_date or date.today()
    return (
        _PRIORITY_WEIGHTS[appointment.prioridade]
        + _TYPE_WEIGHTS[appointment.tipo]
        + _STATUS_WEIGHTS[appointment.status]
        + _deadline_weight(appointment.data, today)
    )


def _build_reason(appointment: Appointment, score: int, reference_date: date) -> str:
    parts: list[str] = []

    if appointment.status == AppointmentStatus.CONCLUIDO:
        parts.append("status concluido reduz a prioridade")
    else:
        parts.append(f"status {appointment.status}")

    parts.append(f"prioridade {appointment.prioridade}")
    parts.append(f"tipo {appointment.tipo}")

    delta_days = (appointment.data - reference_date).days
    if delta_days < 0:
        parts.append("prazo vencido")
    elif delta_days == 0:
        parts.append("prazo hoje")
    elif delta_days == 1:
        parts.append("prazo em 1 dia")
    else:
        parts.append(f"prazo em {delta_days} dias")

    return f"Pontuacao {score}: " + ", ".join(parts) + "."


def build_priority_plan(
    appointments: list[Appointment],
    reference_date: date | None = None,
) -> list[PriorityPlanItem]:
    today = reference_date or date.today()
    prioritized = []

    for appointment in appointments:
        score = calculate_priority_score(appointment, today)
        prioritized.append(
            PriorityPlanItem(
                **appointment.model_dump(),
                pontuacao=score,
                motivo=_build_reason(appointment, score, today),
            )
        )

    return sorted(prioritized, key=lambda item: (-item.pontuacao, item.data, item.titulo))

File: frontend/src/lib/agendaedu/priorities.js
import { PRIORITY_WEIGHTS, STATUS_WEIGHTS, TYPE_WEIGHTS } from "./schema";

function deadlineWeight(appointmentDate, referenceDate) {
  const deadline = new Date(`${appointmentDate}T00:00:00`);
  const reference = new Date(`${referenceDate}T00:00:00`);
  const deltaDays = Math.round((deadline - reference) / 86_400_000);

  if (deltaDays < 0) return 35;
  if (deltaDays === 0) return 40;
  if (deltaDays <= 2) return 35;
  if (deltaDays <= 7) return 25;
  if (deltaDays <= 14) return 15;
  return 5;
}

export function todayIsoDate() {
  return new Date().toISOString().slice(0, 10);
}

export function scoreAppointment(appointment, referenceDate = todayIsoDate()) {
  return (
    PRIORITY_WEIGHTS[appointment.prioridade] +
    TYPE_WEIGHTS[appointment.tipo] +
    STATUS_WEIGHTS[appointment.status] +
    deadlineWeight(appointment.data, referenceDate)
  );
}

export function buildPriorityReason(appointment, score, referenceDate) {
  const deadline = new Date(`${appointment.data}T00:00:00`);
  const reference = new Date(`${referenceDate}T00:00:00`);
  const deltaDays = Math.round((deadline - reference) / 86_400_000);
  const parts = [];

  parts.push(appointment.status === "concluido" ? "status concluido reduz a prioridade" : `status ${appointment.status}`);
  parts.push(`prioridade ${appointment.prioridade}`);
  parts.push(`tipo ${appointment.tipo}`);

  if (deltaDays < 0) {
    parts.push("prazo vencido");
  } else if (deltaDays === 0) {
    parts.push("prazo hoje");
  } else if (deltaDays === 1) {
    parts.push("prazo em 1 dia");
  } else {
    parts.push(`prazo em ${deltaDays} dias`);
  }

  return `Pontuacao ${score}: ${parts.join(", ")}.`;
}
